sql/core/src/test/scala/org/apache/spark/sql/execution/datasources/FileMetadataColumnsSuite.scala

```diff
@@ -0,0 +1,481 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.datasources
+
+import java.io.File
+import java.nio.file.Files
+import java.text.SimpleDateFormat
+
+import org.apache.spark.sql.{AnalysisException, Column, DataFrame, QueryTest, Row}
+import org.apache.spark.sql.functions._
+import org.apache.spark.sql.test.SharedSparkSession
+import org.apache.spark.sql.types.{IntegerType, LongType, StringType, StructField, StructType}
+
+class FileMetadataColumnsSuite extends QueryTest with SharedSparkSession {
+
+  val data0: String =
+    """
+      |jack,24,12345,uom
+      |""".stripMargin
+
+  val data1: String =
+    """
+      |lily,31,,ucb
+      |""".stripMargin
+
+  val schema: StructType = new StructType()
+    .add(StructField("name", StringType))
+    .add(StructField("age", IntegerType))
+    .add(StructField("id", LongType))
+    .add(StructField("university", StringType))
+
+  val schemaWithNameConflicts: StructType = new StructType()
+    .add(StructField("name", StringType))
+    .add(StructField("age", IntegerType))
+    .add(StructField("_metadata.file_size", LongType))
+    .add(StructField("_metadata.FILE_NAME", StringType))
```

Comment:
This looks a bit confusing. IIRC we only have one metadata column: `_metadata`, and `StructField("_metadata.file_size", LongType)` is a column with name "_metadata.file_size", but not getting an inner field `file_size` from a struct-type column `_metadata`

Reply:
yep, that's right: just want to make sure: _metadata.file_size (a field in a struct) and _metadata.file_size (a flat column with a dot in its name) are not the same, we can read all metadata columns properly. Add a comment to explain

See this test: `metadata fields and _metadata.file_size are not the same`

The test and the schema are a little bit "weak", I guess. If you think it is not necessary, we can remove it for sure.